### PR TITLE
Filter out non existent entities in permissions and link shares

### PR DESCRIPTION
**Pull request by meain**
---

#### Does this PR need a docs update or release note?

- [ ] :white_check_mark: Yes, it's included
- [ ] :clock1: Yes, but in a later PR
- [x] :no_entry: No

#### Type of change

- [ ] :sunflower: Feature
- [x] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Supportability/Tests
- [ ] :computer: CI/Deployment
- [ ] :broom: Tech Debt/Cleanup

#### Issue(s)

* #<issue>

#### Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: src/pkg/services/m365/api/groups.go [hunk 1]
```diff
@@ -50,6 +51,25 @@ func (c Groups) GetAll(
 	return getGroups(ctx, errs, service)
 }
 
+// GetAllIDsAndNames retrieves all groups in the tenant and returns them in an idname.Cacher
+func (c Groups) GetAllIDsAndNames(ctx context.Context, errs *fault.Bus) (idname.Cacher, error) {
```
Comment (ryanfkeepers) [inline src/pkg/services/m365/api/groups.go:55]: Request: can you please move this addition to a separate PR?  And, in that change, can you please remove this old func in services favor of this new one? https://github.com/alcionai/corso/blob/main/src/pkg/services/m365/groups.go#L160

File: src/internal/m365/collection/drive/restore_caches.go [hunk 2]
```diff
@@ -21,6 +22,11 @@ type driveInfo struct {
 	rootFolderID string
 }
 
+type AvailableEntities struct {
```
Comment (ryanfkeepers) [inline src/internal/m365/collection/drive/restore_caches.go:25]: Prefer the term `Resource` over `Entity`, please.  Entity is too generic, and can reference non-resource entries.

```suggestion
type ResourceIDNames struct {
```

File: src/internal/m365/collection/drive/permission.go [hunk 3]
```diff
@@ -445,6 +445,102 @@ func UpdateLinkShares(
 	return alreadyDeleted, nil
 }
 
+func filterUnavailableEntitiesInLinkShare(
```
Comment (ryanfkeepers) [inline src/internal/m365/collection/drive/permission.go:448]: these two funcs could be merged pretty easily with generics.

```go
nc filterToExistingResourceEntities[T any](
 	ctx context.Context,
 	perms []T,
 	availableEntities AvailableEntities,
 	oldIDToNewID syncd.MapTo[string],
 ) []T {
```

**Review by ryanfkeepers — APPROVED** [review]
code implementation could be refined, and there's a request for breaking apart the PR.  But nothing blocking.

File: src/internal/m365/collection/drive/permission.go (hunk 3 above)
Comment (meain) [reply src/internal/m365/collection/drive/permission.go:448]: Hmm, maybe I am missing something, but I could not get it folded into one. They are slightly different enough to cause trouble.

File: src/pkg/services/m365/api/groups.go (hunk 1 above)
Comment (meain) [reply src/pkg/services/m365/api/groups.go:55]: Moving this out to a separate PR might get a bit messy as we need it for this PR, but let me create a follow up to remove the old one.

Comment (meain) [reply src/pkg/services/m365/api/groups.go:56]: Created https://github.com/alcionai/corso/pull/4967

File: src/internal/m365/collection/drive/permission.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/permission.go:452]: Ah, looks like you'd need to extend an interface.  Then it'd work.  Take it or leave it, no big deal.

```go
type idAndTyper interface {
  ID() string
  Type() metadata.GV2Type
}

func filterToExistingResourceEntities[T idAndTyper](
 	ctx context.Context,
 	orig []T,
 	availableEntities ResourceIDNames,
 	oldIDToNewID syncd.MapTo[string],
 ) []T {
 	if availableEntities.Users == nil || availableEntities.Groups == nil {
 		logger.Ctx(ctx).Info("no resources provided for filtering")
 		return orig
 	}

 	filtered := []T{}

 	for _, t := range orig {
 		available := false

 		switch t.Type() {
 		case metadata.GV2User:
 			_, ok := availableEntities.Users.NameOf(t.ID())
 			available = available || ok
 		case metadata.GV2Group:
 			_, ok := availableEntities.Groups.NameOf(t.ID())
 			available = available || ok
 		default:
 			// We only know about users and groups
 			// TODO: extend the check to other entity types
 			available = true
 		}

 		if available {
 			filtered = append(filtered, t)
 			continue
 		}

 		// If we have no entities, we can't restore the permission
 		// and so we have to mark it as not restored.
 		oldPermIDToNewID.Store(t.ID(), "")
 	}

 	return filtered
 }
```

Comment (meain) [reply src/internal/m365/collection/drive/permission.go:452]: Yeah that works, but I think I'll skip it the primary reason being that the implementation within is also slightly different. Permissions has a single entry and link shares can have multiple entities. Don't think it will give a big advantage.
